File: Python/LRUCache.py
class LRUCache:

    # @param capacity, an integer
    def __init__(self, capacity):
        self.__cacheMap = {}
        self.__cacheList = None
        self.__cacheListEnd = None
        self.__capacity = max(capacity, 1)

    # @return an integer
    def get(self, key):
        if key not in self.__cacheMap: return -1
        node = self.__cacheMap[key]
        self.__moveToEnd(node)
        return node.v

    # @param key, an integer
    # @param value, an integer
    # @return nothing
    def put(self, key, value):
        if key not in self.__cacheMap:
            self.__addNew(key, value)
        else:
            node = self.__cacheMap[key]
            node.v = value
            self.__moveToEnd(node)
    
    # add a new page
    def __addNew(self, k, v):
        newNode = CacheNode(k, v, None, None)
        if self.__cacheList == None:
            self.__cacheList = newNode;
        else:
            self.__cacheListEnd.ne = newNode
            newNode.pre = self.__cacheListEnd
        self.__cacheListEnd = newNode
        self.__cacheMap[k] = newNode
        if len(self.__cacheMap) > self.__capacity:
            del self.__cacheMap[self.__cacheList.k]
            self.__cacheList.ne.pre = None
            self.__cacheList = self.__cacheList.ne

    # move a page to the end of the list
    def __moveToEnd(self, node):
        if node == self.__cacheListEnd: return
        node.ne.pre = node.pre
        if node == self.__cacheList:
            self.__cacheList = node.ne
        else:
            node.pre.ne = node.ne
        self.__cacheListEnd.ne = node
        node.pre = self.__cacheListEnd
        node.ne = None
        self.__cacheListEnd = node
        
class CacheNode:
    def __init__(self, k, v, pre, ne):
        self.k = k
        self.v = v
        self.pre = pre
        self.ne = ne

File: Go/LRUCache.go
// node in cache list
type CacheNode struct {
    key   int
    value int
    prev  *CacheNode
    next  *CacheNode
}
type LRUCache struct {
    cacheMap       map[int]*CacheNode
    cacheList      *CacheNode
    cacheListEnd   *CacheNode
    capacity       int
}


func Constructor(capacity int) LRUCache {
    return LRUCache {
        cacheMap     : make(map[int]*CacheNode),
        cacheList    : nil,
        cacheListEnd : nil,
        capacity     : capacity,
    }
}

// move a node to the end of the cache list
func (this *LRUCache) moveToEnd (node *CacheNode) {
    if node == this.cacheListEnd {
        return
    }
    node.next.prev = node.prev
    if node == this.cacheList {
        this.cacheList = node.next
    } else {
        node.prev.next = node.next
    }
    this.cacheListEnd.next = node
    node.prev = this.cacheListEnd
    node.next = nil
    this.cacheListEnd = node
}

// add a new cache node to the end of the cache list
func (this *LRUCache) addNew(key int, value int) {
    newNode := CacheNode {
        key : key,
        value : value,
        prev : this.cacheListEnd,
        next : nil,
    }
    if this.cacheList == nil {
        this.cacheList = &newNode
    } else {
        this.cacheListEnd.next = &newNode
    }
    this.cacheListEnd = &newNode
    this.cacheMap[key] = &newNode
    if len(this.cacheMap) > this.capacity {
        delete(this.cacheMap, this.cacheList.key)
        this.cacheList.next.prev = nil
        this.cacheList = this.cacheList.next
    }
}
func (this *LRUCache) Get(key int) int {
    if this.capacity < 1 {
        return -1
    }
    node,exists := this.cacheMap[key]
    if !exists {
        return -1
    }
    this.moveToEnd(node)
    return node.value
}


func (this *LRUCache) Put(key int, value int)  {
    if this.capacity < 1 {
        return
    }
    node,exists := this.cacheMap[key]
    if exists {
        node.value = value
        this.moveToEnd(node)
    } else {
        this.addNew(key, value)
    }
}


/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
